prov/gni/src/gnix_ep.c

```diff
@@ -393,11 +407,165 @@ int gnix_ep_open(struct fid_domain *domain, struct fi_info *info,
 	ep_priv->rx_progress_fn = NULL;
 
 	/*
- 	 * TODO: hookup a nic to this ep, may add more later if
- 	 *  fi_tx_context, etc. is invoked on this endpoing
- 	 */
+	 * TODO: hookup a nic to this ep, may add more later if
+	 *  fi_tx_context, etc. is invoked on this endpoing
+	 */
+
+        list_for_each(&gnix_nic_list,elem,list) {
+		if ((elem->ptag == domain_priv->ptag) &&
+			(elem->cookie == domain_priv->cookie)
+			&& (elem->cdm_id == domain_priv->cdm_id)) {
+			nic = elem;
+			atomic_inc(&nic->ref_cnt);
+			break;
+		}
+	}
+
+	/*
+	 * move this nic to the end of the list, this allows
+	 * for balancing use of nics across ep's for this domain
+	 */
+
+	if (nic) {
+		gnix_list_del_init(&nic->list);
+		list_add_tail(&gnix_nic_list, &nic->list);
+	}
+
+	/*
+	 * no nic found create a cdm and attach
+	 */
+	if (!nic) {
+		nic = calloc(1,sizeof(struct gnix_nic));
+		if (nic == NULL) {
+			ret = -FI_ENOMEM;
+			goto err;
+		}
+
+		/*
+		 * cook up a fake cdm_id, use the 16 LSB of my pid
+		 * with 16 MSBs being obtained from atomic increment of
+		 * a local variable.
+		 */
+		atomic_inc(&gnix_id_counter);
```

Comment:
Is there a convenient place to decrement `gnix_id_counter`?  Seems like it would be nice for error checking/cleanliness purposes to make sure we didn't screw up along the way.